src/internal/easee/signalr.go

```diff
@@ -0,0 +1,451 @@
+package easee
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"sync"
+
+	"github.com/futurehomeno/cliffhanger/root"
+	libsignalr "github.com/philippseith/signalr"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/futurehomeno/edge-easee-adapter/internal/config"
+	"github.com/futurehomeno/edge-easee-adapter/internal/signalr"
+)
+
+const (
+	signalRURI = "/hubs/chargers"
+)
+
+// SignalRManager is the interface for the Easee signalR manager.
+// It manages the signalR connection and the chargers that are connected to it.
+type SignalRManager interface {
+	root.Service
+
+	// Register registers a charger to be managed.
+	Register(chargerID string, cache ObservationCache, callbacks map[ObservationID]func())
+	// Unregister unregisters a charger from being managed.
+	Unregister(chargerID string)
+}
+
+type signalRManager struct {
+	mu       sync.RWMutex
+	client   signalr.Client
+	receiver *SignalRReceiver
+	chargers map[string]chargerItem
+	done     chan struct{}
+}
+
+func NewSignalRManager(client signalr.Client, receiver *SignalRReceiver) SignalRManager {
+	return &signalRManager{
+		client:   client,
+		receiver: receiver,
+		chargers: make(map[string]chargerItem),
+		done:     make(chan struct{}),
+	}
+}
+
+func (m *signalRManager) Start() error {
+	go m.run()
```

Comment:
This is a nitpick and should not make any difference as well as is not per se requirement here from me, but:
Start/stop services should be protected from being started/stopped multiple times. Executing start multiple times will have funny consequences here.

The most proper solution is to either have an idempotent start/stop or erroring when being used in an incompatible phase, depending on the use case.

Reply:
Start/Stop should be called only once in this case, but I'll make it idempotent to be safe. Thanks for the hint!